### PR TITLE
Mapsforge themes compatibility

### PR DESCRIPTION
**Pull request by Longri**
XmlMapsforgeThemeBuilder.java accept a Mapsforge-Theme!

There are some "TODO's" in this file!

### CODE REVIEW

**Comment by Longri** [comment]
Please leave it open, I'm working on the line dash-array, at the moment!

**Comment by devemux86** [comment]
Sure, take your time and can inform when something is completed!

Anyway I have to find also the time to properly review / test it. :slightly_smiling_face:

**Comment by Longri** [comment]
Now I will try to solve 'Caption'
   - unknown attribute in element caption () : 0 = k
   - unknown attribute in element caption () : 1 = font-style
   - unknown attribute in element caption () : 2 = font-size
   - unknown attribute in element caption () : 3 = fill
   - unknown attribute in element caption () : 4 = stroke
   - unknown attribute in element caption () : 5 = stroke-width

**Comment by devemux86** [comment]
Nice! Some quick comments:
- Why the need of `AtomicBoolean`?
- For the Mapsforge [position](https://github.com/mapsforge/mapsforge/blob/master/resources/renderTheme-v6.xsd#L13) tag let's use the [default](https://github.com/mapsforge/mapsforge/blob/master/docs/Rendertheme.md#symbol-positioning) below option with VTM `dy`, until we implement separately #195.
- The check for `isMapsforgeTheme` can be hosted in a static method at new `org.oscim.theme.ThemeUtils` class.

**Comment by devemux86** [comment]
> Now I will try to solve 'Caption'

All fields exist renamed in VTM, don't they?

**Comment by Longri** [comment]
> > Now I will try to solve 'Caption'

> All fields exist renamed in VTM, don't they?

I saw they, logged here although there was no mistake. 
Fixup comes with the next commit.



> The check for isMapsforgeTheme can be hosted in a static method at new org.oscim.theme.ThemeUtils class.

I will move this code to org.oscim.theme.ThemeUtils



> Why the need of AtomicBoolean?

On this point I need a final object that I can modify!
Do you have another idea?


> For the Mapsforge position tag let's use the default below option with VTM dy, until we implement separately #195.

Ok, I will dont touch this, first!

**Comment by devemux86** [comment]
> Do you have another idea?

If having that code in a separate static boolean method, can use a simple `boolean` field and don't need also any special `SAXTerminatorException`, just return true or false as soon as the parse decides the outcome.

**Comment by Longri** [comment]
> If having that code in a separate static boolean method, can use a simple boolean field and don't need also any special SAXTerminatorException, just return true or false as soon as the parse decides the outcome.

Nice! I will do that!

**Comment by Longri** [comment]
> The check for isMapsforgeTheme can be hosted in a static method at new org.oscim.theme.ThemeUtils class.

org.oscim.theme.ThemeUtils does not exist, should I create them?

**Comment by devemux86** [comment]
> org.oscim.theme.ThemeUtils does not exist, should I create them?

Yes, it would be useful!

**Comment by Longri** [comment]
I have add the ThemeUtils.java!

But with my version of 'isMapsforgeTheme()'
I didn't get what you meant.
Maybe you could change that, please.



> I think in Mapsforge the "repeat-start" is only for the 1st symbol at start of line, not for each symbol along the line?
I will explain with a graphic, give me a moment!

**Comment by Longri** [comment]
I think the point has now been reached where the Themer are called for.

We could send them a version of the ThemeComparator and wait to see what errors they find in the representation!
 
Maybe you can merge now, after your review?

**Comment by devemux86** [comment]
> Maybe you can merge now, after your review?

Thanks, I'll review it!

File: vtm/src/org/oscim/theme/XmlMapsforgeThemeBuilder.java [hunk 1]
```diff
@@ -0,0 +1,1226 @@
...
+                String val = value.toUpperCase(Locale.ENGLISH);
+                if ("WAY".equals(val))
+                    element = Rule.Element.WAY;
+                else if ("NODE".equals(val))
+                    element = Rule.Element.NODE;
+            } else if ("k".equals(name)) {
+                if (!"*".equals(value))
+                    keys = value;
+            } else if ("v".equals(name)) {
+                if (!"*".equals(value))
```
Comment (devemux86) [inline vtm/src/org/oscim/theme/XmlMapsforgeThemeBuilder.java]: Why we need the wildcard checks for key and value?

**Comment by Longri** [comment]
I don't know! I have copied the XmlThemeBuilder.java and then modified only this what I need!

**Comment by devemux86** [comment]
Really nice work!

Merged via b695d43fee05671b096e588e612dd6b94fada91c with improvements via f4f8eb8d1cbbf96b159a9102806aa3bf3dde17e5.

BTW I added the "line symbol" and "dash array" code also in VTM theme parser, to test it there.

**Comment by devemux86** [comment]
There are some render problems with line symbol e.g. `oneway`, but let's discuss them in their issues.

**Comment by mathieudebrito** [comment]
Hi Guys, 

I just ran into this issue : 
```
09-05 11:01:45.133 13818-13818/? W/System.err: org.oscim.theme.IRenderTheme$ThemeException: stream closed
09-05 11:01:45.133 13818-13818/? W/System.err:     at org.oscim.theme.XmlThemeBuilder.read(XmlThemeBuilder.java:114)
09-05 11:01:45.134 13818-13818/? W/System.err:     at org.oscim.theme.ThemeLoader.load(ThemeLoader.java:54)
09-05 11:01:45.134 13818-13818/? W/System.err:     at org.oscim.theme.ThemeLoader.load(ThemeLoader.java:46)
09-05 11:01:45.134 13818-13818/? W/System.err:     at org.oscim.map.Map.setTheme(Map.java:203)
09-05 11:01:45.134 13818-13818/? W/System.err:     at org.oscim.map.Map.setTheme(Map.java:195)
```
It seems related to these updates, since our map code didn't change.
Do you guys have any idea ? (yep, it's minimal, but that's all I got :p)

**Comment by Longri** [comment]
is it a Mapsforge Theme?
Then you must update to the latest SNAPSHOT!
The error messages says that you try to load a Theme with 'XmlThemeBuilder'!
For Mapsforge we have 'XmlMapsforgeThemeBuilder', but is not called!

**Comment by devemux86** [comment]
Is it reproducible, trying to open an external VTM theme xml file?

**Comment by mathieudebrito** [comment]
Hi @Longri @devemux86 , thanks for your very quick answer !
It's a mapzen theme. The version I use is master-SNAPSHOT.

I retrieve the theme like this (in order to filter the categories)
```
    StreamRenderTheme theme = new StreamRenderTheme("", CustomVtmThemes.class.getResourceAsStream("/assets/renderthemes/my_awesome_theme.xml"), new XmlRenderThemeMenuCallback() {
            @Override
            public Set<String> getCategories(XmlRenderThemeStyleMenu style) {
            ...
```

The issue seems reproducible since I tried with the default mapzen theme (from this repository) and ran into the same issue :p

Cheers

**Comment by devemux86** [comment]
Also [MapsforgeStyleActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/MapsforgeStyleActivity.java) loading a theme from `assets` folder fails to run now.
We have to fix it!

**Comment by devemux86** [comment]
@mathieudebrito thanks for the report!

**Pull request closed without merging**